include/real/exact_number.hpp

```diff
@@ -693,190 +694,322 @@ namespace boost {
+             *  @param:  upper: if true: error lies in [0, +epsilon]
+             *                  else: error lies in [-epsilon, 0], here epsilon = 1*base^(-maximum_precision)
+             *  @author: Kishan Shukla
+             */
+            void divide_vector(const exact_number<T> divisor, unsigned int maximum_precision, bool upper) {
+
+                newton_raphson_division(divisor, maximum_precision, upper);
+
+            }
+
+            /*          BINARY SEARCH BASED DIVISION
+             *      an approximate division method, used in the initial guess of reciprocal in Newton Raphson
+             */
+
+            void binary_search_division(
+                const exact_number<T> divisor,
+                unsigned int maximum_precision){
+
+                /* Exceptions */
                 if (maximum_precision > (unsigned int) std::abs(std::numeric_limits<exponent_t>::min())) {
                     throw exponent_overflow_exception();
                 }
 
-                min_boundary_n.digits = {1};
-                min_boundary_n.exponent = -1 * (maximum_precision);
-                min_boundary_n.positive = false;
-
-                min_boundary_p.digits = {1};
-                min_boundary_p.exponent = -1 * (maximum_precision);
-
-                T base = (std::numeric_limits<T>::max() /4)*2 - 1;
-                T H = base/2 + 1;
-                half.digits = {H};
-                half.exponent = 0;
+                exact_number<T> zero = exact_number<T> ();
+                if(divisor == zero){
+                    throw divide_by_zero();
+                }
 
-                // we ignore exponents, then set them in the end.
-                // (a * 10^x) / (b*10^y) = (a*10 / b*10) * 10^((x-1)-(y-1))
-                // 100/20 -> 1/2 * (10)
-                int exponent_dif = (numerator.exponent - 1) - (divisor.exponent - 1);
+                /* Special Cases */
+                if((*this) == zero){
+                    return;
+                }
 
-                numerator.exponent = 1;
-                divisor.exponent = 1;
-                tmp.digits = {1};
-                tmp.exponent = 1;
+                exact_number<T> one;
+                one.digits={1};
+                one.exponent=1;
+                if(divisor == one){
+                    return;
+                }
 
-                exact_number<T> zero = exact_number<T>(); 
+                bool positive = (this->positive == divisor.positive);
 
-                if (divisor == tmp) {
-                    this->exponent = exponent_dif + 1;
+                exact_number<T> minus_one = one;
+                minus_one.positive = false;
+                if(divisor == minus_one){
                     this->positive = positive;
                     return;
                 }
 
-                if (divisor == (*this)) { 
-                    (*this) = tmp;
-                    (*this).positive = positive;
+                if(divisor == (*this)){
```

Comment:
Not sure this is a good idea, you going to do a full comparizon of the number for optimizing a case that is not expected to be that common, it may be ok to compare the addresses of the objects with is immediate and optimize just for the case they are both referencing the same number, but not doing a compare of all digits.

Reply:
Yes, you are right. For large numbers, this check would be very expensive and also this is a rare case. Comparing the complexity of the whole algorithm, I thought this check wouldn't affect much the complexity as a whole. I'll remove it.

Reply:
I also agree with damian. Comparing the addresses seems a good solution that optimizes for those cases without worsen all other cases.